### PR TITLE
nfs: update operator for spec validation

### LINKED ISSUE #2036: Validate accessibility permissions for NFS volumes
**Is this a bug report or feature request?**
* Feature Request

**What should the feature do:**
Currently, the Rook NFS operator has no validation for `exports.server.accessMode` parameter.
This feature should validate `ReadWrite`, `ReadOnly` and `none` options as mentioned in the [NFS CRD design](https://github.com/rook/rook/blob/master/design/nfs.md#nfs-crd) as the supported parameters and should return an error if something other than this is entered.

**What is use case behind this feature:**
This feature will validate access permissions to the NFS volume created.

### PR DESCRIPTION
**Pull request by tsullens**
adding validation of NFSServerSpec in accordance
with issues #2036 and #2037

Signed-off-by: tsullens <tsullens@mix.wvu.edu>

**Description of your changes:**
Added validation functions for the `exports.server.accessMode` and `exports.server.squash` NFS CRD Parameters as outlined [here](https://github.com/rook/rook/blob/master/design/nfs.md#nfs-crd).

**Which issue is resolved by this Pull Request:**
Resolves #2036, #2037 

**Checklist:**
- [ ] Documentation has been updated, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.
- [ ] Comments have been added or updated based on the standards set in [CONTRIBUTING.md](../blob/master/CONTRIBUTING.md#comments)

[skip ci]

### CODE REVIEW

File: pkg/operator/nfs/controller.go [hunk 1]
```diff
@@ -424,3 +429,37 @@ func (c *Controller) onDelete(obj interface{}) {
...
+	for _, export := range serverConfig {
+		if err := validateAccessMode(export.Server.AccessMode); err != nil {
+			return err
+		}
+		if err := validateSquashMode(export.Server.Squash); err != nil {
+			return err
+		}
+	}
+	return nil
+}
```
Comment (rohan47) [inline pkg/operator/nfs/controller.go]: A blank line after this will be great.

File: pkg/operator/nfs/controller.go [hunk 2]
```diff
@@ -424,3 +429,37 @@ func (c *Controller) onDelete(obj interface{}) {
...
+func validateAccessMode(mode string) error {
+	switch mode {
+	case "ReadOnly":
+	case "ReadWrite":
+	case "none":
+	default:
+		return fmt.Errorf("Invalid value (%s) for accessMode, valid values are (ReadOnly, ReadWrite, none)", mode)
+	}
+	return nil
+}
```
Comment (rohan47) [inline pkg/operator/nfs/controller.go]: a blank line after this

File: pkg/operator/nfs/controller.go (hunk 1 above)
Comment (tsullens) [reply pkg/operator/nfs/controller.go]: Got it.

**Comment by tsullens** [comment]
@rohan47 I think I can handle the unit and integration tests for this, it will be a few days before I am able to work on it again though.

**Comment by rohan47** [comment]
@tsullens Thanks :smiley:

File: pkg/operator/nfs/controller.go [hunk 3]
```diff
@@ -424,3 +429,37 @@ func (c *Controller) onDelete(obj interface{}) {
...
+		}
+		if err := validateSquashMode(export.Server.Squash); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+func validateAccessMode(mode string) error {
+	switch mode {
+	case "ReadOnly":
```
Comment (jbw976) [inline pkg/operator/nfs/controller.go]: does the underlying NFS server config allow different casing here? e.g., would `readonly` be OK?  if so, we need to normalize casing here for this check.

**Comment by jbw976** [comment]
I think just unit tests for these functions would be acceptable for this, I don't think any new full integration tests are needed.

**Comment by tsullens** [comment]
After taking a further look I'm agreed with @jbw976. I've added a unit test for `validateNFSServerSpec` to cover these existing cases and any other cases that may be added for validation.

File: pkg/operator/nfs/controller.go (hunk 3 above)
Comment (tsullens) [reply pkg/operator/nfs/controller.go]: OK so after getting the time to dive deeper into this, the [nfs-ganesha export config](https://github.com/nfs-ganesha/nfs-ganesha/blob/master/src/config_samples/config.txt) has squash options all in lowercase which I guess we will want to conform to. 
AccessMode/AccessType is slightly different, where nfs-ganesha lists e.g. `None`, `RW`, etc. This highlights maybe another necessity, as we would need to transform the CRD to the correct values for the ganesha export and this is already started in the `createGaneshaExport` func however this is only handling `ReadOnly` -> `RW` and not covering the other AccessTypes which will need to be done eventually... 

So kinda what I am thinking is for the validation step to ignore case for these values via my suggestion above, and handle transformations for `AccessMode` and `Squash` in `createGaneshaExport`. Does that works for everyone @jbw976 @rohan47? I am totally open to comments/suggestions here.

**Comment by tsullens** [comment]
OK @rohan47, I want to say this is about done but I think I managed to mess up the commit history and CI does not want to build...  
I've pulled everything into a separate branch that has a clean history if you'd like me to push that (or I guess I could just `--force` on this branch), but I'm not sure what the deal is with CI since tests work for me and I can't reproduce the problem locally... perhaps I'm missing something there.

**Comment by rohan47** [comment]
@tsullens regarding the commit history I am not sure what will be a good option. 
and CI failed because of the commit history mess up as I can see nfs server test passed and `github.com/rook/rook/pkg/operator/ceph/provisioner/controller`

**Comment by rohan47** [comment]
`TestBlockMountUnMountSuite` failed.
`TestNfsSuite` passed
maybe re running the build would fix it.

**Comment by travisn** [comment]
@tsullens @rohan47 i restarted the build. it does look like one of our known random issues.

**Comment by tsullens** [comment]
Thanks @travisn, but it looks like it failed again at pretty much the same place.

**Comment by travisn** [comment]
the CI is being extra difficult this week. i think we'll need to skip the CI to get this merged while i look into the CI issues independently. @jbw976 anything else to approve this PR?

**Pull request merged**